### ISSUE TITLE
Request: Enabling Simultaneaous Controller+Keyboard/Mouse Input (Mixed Input) across all RE Engine-based games

### ISSUE THREAD

**Issue by AL2009man** [praydog/REFramework#674]
To explain a bit of context: Simultaneously Controller+Keyboard/Mouse, **Mixed Input (or Hybrid Input for newer Ubisoft games) for short**, is a Input Method that enables users to use the Gamepad and Keyboard/Mouse Inputs at the same time. Typically: this is used for something like Gyro Aiming (most Controllers like PlayStation and Nintendo Switch), Tracking Aiming (for Steam Controller or Steam Deck, in combination with Gyro or not), Gaming Keypads (Azeron, Razer), Analog Keyboard (Wooting Keyboard), physically using Gamepad and Mouse device, etc. You could say this is an Accessibility-related request.

**for simplification's sake: I will be referring to "Mixed Input" going forward.**

Normally, this method will/may introduce issues such as:

- **Controller/KB+M Button prompt flickers.**
- **Mouse and Joystick Camera Input fights with each other**.
- **Locks you to One Input Method**
- Joystick Camera Input starts to behave erratically.
- Certain gameplay/UI elements (example: key gameplay functions like Sprinting and Interacting is broken in Metro Exodus, KB/M and Controller-specific UI behaves erratically in Sunset Overdrive)
- Input Method's entire Aiming mechanic changes (in the case with Resident Evil 5)

quick off-topic: It's rare to find a game that allows you to use Mixed Input without conflictions (or as an option in the case with CULTIC and Death Stranding via Button Prompt Icon settings), which is why Valve Corporations recommends to not tie KB/M Button prompt switch based on Mouse Movement when preparing for "Steam Deck support". **This is one of the most common chief compatibility issues that faced Mixed Input supports.** But if you'd like to learn more: [I suggest reading this article by Flick Stick creator and current-gyro expert Jibb Smart on Mixed Input](https://cohost.org/jibbsmart/post/435743-1-million-reasons-to).

As of this writing: Here's the list of known RE Engine-based games to be confirmed (by yours truly) to allow Mixed Input:

- Devil May Cry 5
- Monster Hunter RISE
- Resident Evil 4 Remake
- Exoprimal

basically: all RE Engine-based games _after_ **Monster Hunter RISE** (aka starting in that engine branch) will support Mixed Input, aside of Button prompt flickering.

You may have noticed that RE4 Remake is the only Resident Evil game to allow that, but not...the rest of RE Engine-based Resident Evil games? Well, this is because it suffers from the most nefarious of issues from the list [**highlighted on the bold text**] where it outright prevents you from using both Inputs at the same time. You are restricted to only one Input Method.

until Resident Evil 4 Remake: Resident Evil games always never allowed you to use Mixed Input, but given the context of RE4 Remake allowing you to use Mixed Input: I wonder if there's a way to untick the "Input Switch Lock" within the config/hex/REFramework that allows players to fully remove that restriction?

this option could benefit those who like to use Mixed Input on various setups in all RE Engine-based games instead of a small few.

**Comment by LAN021** [comment]
I would very much like this as well, and possibly an option for locking the prompts to either controller or KB/M, regardless of input type. This is available in Death Stranding, as noted, and is extremely useful. If it's possible to lock the prompts in the four games with support for mixed input, that would be great as well (especially MHR).

On PCGamingWiki, Resident Evil 3 and Village are marked as having support for mixed input. Is this incorrect or does it not work properly? It's called "Simultaneous controller+KB/M" there, in the Input section.

**Comment by WiFiAdapter** [comment]
Second this, can't properly play RE8 because of the lack of mixed input.

**Comment by AL2009man** [comment]
> I would very much like this as well, and possibly an option for locking the prompts to either controller or KB/M, regardless of input type. This is available in Death Stranding, as noted, and is extremely useful. If it's possible to lock the prompts in the four games with support for mixed input, that would be great as well (especially MHR).
> 
> On PCGamingWiki, Resident Evil 3 and Village are marked as having support for mixed input. Is this incorrect or does it not work properly? It's called "Simultaneous controller+KB/M" there, in the Input box, under "Additional information".

As noted in the OP: they got that information wrong, and you can confirm that if you download the demos (or own the game). I really get the feeling they only swap between two input methods but not simultaneously.

**Comment by LAN021** [comment]
> > On PCGamingWiki, Resident Evil 3 and Village are marked as having support for mixed input. Is this incorrect or does it not work properly? It's called "Simultaneous controller+KB/M" there, in the Input box, under "Additional information".
> 
> As noted in the OP: **nearly all the Resident Evil games I've played thus far historically never supports Mixed Input (prior to Resident Evil 4 Remake), both RE3 Remake and RE Village is apart of it**. they got that information wrong, and you can confirm that if you download the demos (or own the game). I really get the feeling they only swap between two input methods but not simultaneously.

Ok, was just looking for confirmation of that. I don't own the games, but if it's possible to get these features somehow, including locked prompts, I have more reasons to get them in the future, especially Monster Hunter Rise, which I can't play at all because of the control scheme (I use mixed input for remapping).

**Comment by WiFiAdapter** [comment]
I got it working. Now I only need to make a really simple script.
The only thing I changed was the following: app.HIDManager -> AutoGenerated Types -> 0x58: app.HIDInputMode.Value -> app.HIDInputMode.Value -> Reflection Properties -> s32: 0

**Comment by LAN021** [comment]
> I got it working. Now I only need to make a really simple script. The only thing I changed was the following: app.HIDManager -> AutoGenerated Types -> 0x58: app.HIDInputMode.Value -> app.HIDInputMode.Value -> Reflection Properties -> s32: 0

Great work. How do the prompts work after that, do they change based on input? Is it possible to lock them to one type?

**Comment by AL2009man** [comment]
> I got it working. Now I only need to make a really simple script. The only thing I changed was the following: app.HIDManager -> AutoGenerated Types -> 0x58: app.HIDInputMode.Value -> app.HIDInputMode.Value -> Reflection Properties -> s32: 0

That's great! I wonder how we are going to approach this? 

some of RE Engine-based games already allowed for Mixed Input (Resident Evil 4 Remake is a recent example) but the KB/M glyph switch is based on Mouse Movement. I assume a toggle will give users the ability to forcefully lock Controller or KB/M Glyphs instead of automatic (or change the KB/M prompt switch as a keybind/mouse press/click as per Valve's recommendations). 

the rest (like Resident Evil 7 the DX12 edition, Resident Evil 2/3 Remakes and Resident Evil Villages)? You're forced to pick one Input Method, so I wonder how these titles will fare with the script?

**Comment by WiFiAdapter** [comment]
The glyphs are locked to pc symbols which could be a little bit troubling on the steam deck. 
The problem right now is that it seems that changing reflection properties as I did is not a good way to implement stuff. I’d need to find a method or maybe the actual method that is called upon using the controller. 
I will make another comment should I find something out about it.

**Comment by LAN021** [comment]
@WiFiAdapter It's still great progress if you have managed to enable mixed input with locked KB/M prompts. It would be worth adding it to REFramework like that and add locked controller prompts later if you find out how to implement it.

What happens when you use controller input? Is there no switching? That means the game probably prioritizes KB/M prompts in case of input registered at the same time. Some games prioritize controller prompts and some KB/M prompts. Have you tried setting a controller input to be always on in Steam Input, for instance, and see how it works? Or just holding down a keyboard key and moving a stick or pressing a button on the controller? That usually reveals what the game prioritizes.

**Comment by dglt1** [comment]
> The glyphs are locked to pc symbols which could be a little bit troubling on the steam deck. The problem right now is that it seems that changing reflection properties as I did is not a good way to implement stuff. I’d need to find a method or maybe the actual method that is called upon using the controller. I will make another comment should I find something out about it.
> 
> Edit: I managed to make a script that allows mixed input, but I wasn't able to change the glyphs and honestly I don't know where to look for the right variables or methods.

hi, would you mind sharing the mixed input script you made?

**Comment by Vaizarr** [comment]
This is exactly what I'm trying to fix! You can add Chivalry2 to the list. I'm using an Azeron keypad and because it keeps rapidly switching between inputs (controller/kb/mouse) it causes rapid stuttering. I've spent hours trying to work around it. Can you share some information how you solved some of the issues? How did you change these settings you're talking about?

**Comment by AL2009man** [comment]
> > The glyphs are locked to pc symbols which could be a little bit troubling on the steam deck. The problem right now is that it seems that changing reflection properties as I did is not a good way to implement stuff. I’d need to find a method or maybe the actual method that is called upon using the controller. I will make another comment should I find something out about it.
> > Edit: I managed to make a script that allows mixed input, but I wasn't able to change the glyphs and honestly I don't know where to look for the right variables or methods.
> 
> hi, would you mind sharing the mixed input script you made?

Yeah, it's been 3 months since this was last reported. might be a good time to share the incomplete code to everyone. Someone else might look into it using your work as a basis. @LAN021 



> This is exactly what I'm trying to fix! You can add Chivalry2 to the list. I'm using an Azeron keypad and because it keeps rapidly switching between inputs (controller/kb/mouse) it causes rapid stuttering. I've spent hours trying to work around it. Can you share some information how you solved some of the issues? How did you change these settings you're talking about?

Um...Chivalry 2 uses **Unreal Engine 4** and not **RE Engine**. It's a bit redundant to post it here, but I suggest posting it to official game's forum sites.

...however: the vast majority of Unreal Engine 4-based games play nicely with both input methods aside of Button prompt flickering. **It's rare to find a game that doesn't have a problem with button prompt flicker**.

although: There are some games (like Fortnite) that intentionally went out of their way to prevent some players from using that specific setup (like your Azeron keypad), so I wouldn't be surprised if Chivalry 2 is one of the rare UE4 games that prevents it.

**Comment by LAN021** [comment]
@AL2009man I don't know if you mentioned me because you thought I made the script, but it was actually @WiFiAdapter. I'm awaiting this as eagerly as you are, it would allow me to play MHR and the RE Engine Resident Evil games without any issues. If this can't be achieved, then maybe someone can find a way to remap controller inputs, at least swapping the bumpers/triggers as in RE7.

**Comment by AL2009man** [comment]
> @AL2009man I don't know if you mentioned me because you thought I made the script, but it was actually @WiFiAdapter. I'm awaiting this as eagerly as you are, it would allow me to play MHR and the RE Engine Resident Evil games without any issues. If this can't be achieved, then maybe someone can find a way to remap controller inputs, at least swapping the bumpers/triggers as in RE7.

Ops, I accidentally pinged the wrong person. Sorry! 

but still, I'm already expecting far more RE Engine-based games to support Mixed Input going forward.

**Comment by AL2009man** [comment]
> The glyphs are locked to pc symbols which could be a little bit troubling on the steam deck. The problem right now is that it seems that changing reflection properties as I did is not a good way to implement stuff. I’d need to find a method or maybe the actual method that is called upon using the controller. I will make another comment should I find something out about it.
> 
> Edit: I managed to make a script that allows mixed input, but I wasn't able to change the glyphs and honestly I don't know where to look for the right variables or methods.

Hey, quick update: so apparently...since the **Sunbreak update (or when they implemented RawMouse-- i dunno the timeframe)**: Monster Hunter RISE **no longer cause button prompt to flicker based on Mouse Movement**, It now only switches to Keyboard/Mouse prompts when you press a Keyboard/Mouse key. Aside of Camera conflict between XInput Right Stick Camera and Mouse Input : This will make Mixed Input far easier to use on a Steam Deck.

If you already having problems looking into the right variables/methods on locking the button prompts since you last worked on the script: I might suggest looking at how Monster Hunter RISE (since Sunbreak update) as a point of reference on how they handle button prompt flickering. 

(Even if you don't have Monster Hunter RISE: you can download the demo, as it's also been updated to the Sunbreak expansion)

**Comment by LAN021** [comment]
I was really anticipating this but @WiFiAdapter seems to be inactive, so I'm wondering if there is anyone that can continue this, as @AL2009man suggested in the post before.

> I got it working. Now I only need to make a really simple script. The only thing I changed was the following: app.HIDManager -> AutoGenerated Types -> 0x58: app.HIDInputMode.Value -> app.HIDInputMode.Value -> Reflection Properties -> s32: 0

Can anyone work with this?

**Comment by juliohintze** [comment]
Based on @WiFiAdapter solution, I made a script that works.

Be aware: I know next to nothing about Lua and REFramework. WifiAdapter himself said that changing Reflection properties like this isn't the best idea. I don't know why, but I believe him. This script might screw other things over, and I'd probably have no idea how to fix it.

But for those who are interested:

```lua
local game_name = reframework:get_game_name()
local is_re8 = game_name == "re8"

if not is_re8 then
  error("Unsupported game: " .. game_name)
end

local function onUpdate()
  hid_manager = sdk.get_managed_singleton(sdk.game_namespace("HIDManager"))
  inputMode = hid_manager:call("get_inputMode")

  if inputMode.V == 1 then
    inputMode.V = 0
  end
end

sdk.hook(
  sdk.find_type_definition(
    sdk.game_namespace("HIDManager")
  ):get_method("doUpdate"),
  nil,
  onUpdate
)

```

Just drop a MixedInputs.lua inside the autorun folder and you're golden.
If anyone have an idea how to fix the glyphs, it would be nice.

**Comment by LAN021** [comment]
That's outstanding, thank you for your work. How do the prompts work after this? Do they auto-switch based on input? Is one or there other prioritized?

**Comment by juliohintze** [comment]
It's exactly as WifiAdapter said. It does let you use mixed inputs, but the side effect is that the prompts are locked to KB/M.
I'll try looking into that later, but I don't even know where to start.

Having no experience with REFramework (or modding at all) made this task a LOT harder. If it wasn't for @WiFiAdapter's tip, I wouldn't be able to do it.

Also: I tried playing with gyro aiming yesterday, and the game has this auto-aim that REALLY makes things weird.
I don't think the game lets you turn that off. That's another thing to look into.

**Comment by LAN021** [comment]
Yeah, this is still outstanding, even if you're not experienced with this particular thing.

As I said previously:
> It's still great progress if you have managed to enable mixed input with locked KB/M prompts. It would be worth adding it to REFramework like that and add locked controller prompts later if you find out how to implement it.

How does it work in different RE engine games?

**Comment by LAN021** [comment]
Would be much appreciated, and just let us know if you make any other progress, such as finding a way to change prompts to controller. Maybe there's even a way to force PlayStation prompts in games that have them built in so you don't need to use texture mods.

**Comment by AL2009man** [comment]
> I put an if statement to make sure that the script will only run in RE8. I can change that later, but I don't know if it'll work for other RE Engine games.
> 
> I can test it for RE2 and RE3, though. I don't think it would be that hard, but you never know.

I can confirm that the script does work on Resident Evil Village, but doesn't work on Resident Evil 2 Remake, even if I replaced `is_re8 = game_name == "re8"` to `is_re2 = game_name == "re2"`

**Comment by juliohintze** [comment]
Yeah, I just looked into it. I have no idea how to make a script like this for other games, or even how to fix the prompts for RE8.
Maybe it could be done using Cheat Engine, tho.

**Comment by grayespinoza** [comment]
Hey all! Late response, but I'm worried that the **lack of mixed input support might be hard-coded**. I adapted what we have so far into RE2 and RE3
```lua
local game_name = reframework:get_game_name()

local function onUpdate()
  -- Input Device Manager for RE2
  if game_name == "re2" then
    input_system = sdk.get_managed_singleton("app.ropeway.InputSystem")
  end

  -- Input Device Manager for RE3
  if game_name == "re3" then
    input_system = sdk.get_managed_singleton("offline.InputSystem")
  end

  local inputMode = input_system:call("get_InputMode")
  if inputMode ~= 0 then
    input_system:call("set_InputMode", 0)
  end
end

if game_name == "re2" then
  sdk.hook(
    sdk.find_type_definition("app.ropeway.InputSystem"):get_method("update"),
    nil,
    onUpdate
  )
end

if game_name == "re3" then
  sdk.hook(
    sdk.find_type_definition("offline.InputSystem"):get_method("update"),
    nil,
    onUpdate
  )
end
```
and instead of switching inputs, it just disables the K&M. Though, I'm not entirely convinced that
```lua
if inputMode ~= 0 then
  input_system:call("set_InputMode", 0)
end
```
should allow for mixed input. Does anyone actually know why this should be the case in RE8? Is it a different implementation of the method, because it seems like this new behavior is actually valid? Anywhomst, I'll continue trying, but I'm leaving soon to a summer program, so I won't be able to dedicate much of any time until after summer. I hope we can get mixed input and gyro aiming soon for the non-supported RE Engine games!

**Comment by grayespinoza** [comment]
> > I put an if statement to make sure that the script will only run in RE8. I can change that later, but I don't know if it'll work for other RE Engine games.
> > I can test it for RE2 and RE3, though. I don't think it would be that hard, but you never know.
> 
> I can confirm that the script does work on Resident Evil Village.
> 
> **But doesn't work on Resident Evil 2 Remake**, even if I replaced `is_re8 = game_name == "re8"` to `is_re2 = game_name == "re2"`
> 
> ![Screenshot 2024-04-05 141722](https://github.com/praydog/REFramework/assets/67606569/b5b37f9e-7375-4eee-b538-d91846ceab77)

To add more information to this (so anyone trying can learn), the reason switching the `game_name` didn't work is that the manager for input devices differs from RE2 to RE8. In RE8, the manager was `HIDManager`, but in RE2 the manager is `InputSystem`. Furthermore, in RE2 there's some intermediary class, `ropeway`, between the namespace and `InputSystem` which is why we can't use `sdk.game_namespace()` to simplify the implementation between RE2 and RE3.

**Comment by LAN021** [comment]
@grayespinoza That's unfortunate, but thank you for looking into this. Do you happen to know if it's possible to lock prompts to one type in RE Engine games that support mixed input?

**Comment by grayespinoza** [comment]
@LAN021 It's funny that you ask, because I was trying to implement this myself today! I wanted to lock the controller prompts to PlayStation icons. It didn't work out, but I'm certain you're able to! I'm pretty sure I've seen people have this behavior in Monster Hunter Rise before—without just replacing all the controller icon textures. I'll ping you if I write the script, but honestly don't expect it too soon as this is my number one priority in terms of modding.

**Comment by LAN021** [comment]
@grayespinoza Ok, that's greatly appreciated. I would like it in RE4R and Monster Hunter Rise/Wilds especially.

**Comment by grayespinoza** [comment]
I have some notations and an update.

1. `input_system:call("set_InputMode", 0)` is a hard lock on input so that's why it never reads K&M
2. namespaces aren't just `app` and `offline`, `app.ropeway` is also a namespace, so `sdk.game_namespace()` works perfectly fine

Two days ago, I got RE7 to work, so now we have **RE7 and RE8 with mixed input**.
[mixed_input.lua](https://github.com/grayespinoza/REScripts/blob/main/mixed_input.lua)

The way these two work is (from what I understand) we're telling the game to read controller input as K&M input. A side-effect of this is that, the right analog ceases to work; I'm not too sure why though, as the left analog is fine and it doesn't seem to be emulating WASD. I tried looking for methods or fields that read input to the right analog so that I can re-enable it by only reading controller input as K&M while right analog isn't triggered, but I couldn't find anything.

Futher, I think the reason we were having problem initially getting it to other games is that RE2,3,7 don't "auto generate" `Input Mode`, but I was able to get around this by accessing fields in RE7—this did not work however with RE2 and RE3.

Lastly, as essentially we don't have "controller input", the input icons are always K&M, and I haven't found any system that renders or calls specific icons—I'm worried its tied to `Input Mode`. With this in mind, if we were to somehow find out how to seperately call icons, the games themselves would need to natively support (most popular gyro) DualSense, which RE2,3 don't, but it seems RE7,8 might now on PC? It might be better to replace the icons with a texture mod instead and spend time elsewhere.

I'll continue with the updates in case another developer get's interested.

**Comment by grayespinoza** [comment]
I have a notation and an update.

1. RE7 and RE8 do not have extra controller icon support, that's only available in RE4

In RE7 I got the right analog back working by exactly

> only reading controller input as K&M while right analog isn't triggered

so now you can look around as usual with the right analog—and now there's icon flickering. Though, I don't know if it's interesting but if you stay in controller mode and use your gyro without triggering the switch to K&M, the prompts will remain as controller prompts, not K&M prompts, i.e.,

1. Use your right analog so icons switch to controller
2. Activate gyro
3. The prompts should've stayed as controller, but the gyro is clearly emulating a mouse

However, in honesty, after writing this portion, I don't think it's anything meaningful, I don't think this is a lead as I looked thoroughly through RE7 and RE8 for controllers/managers of icon prompts, but I saw nothing. Furthermore, this exact method did not work in RE8 as I also think there might be a hard lock on switching inputs that's doing this, but I can't be sure.

As always, I'll have an update if I get meaningful functionality or notations, and an updated script will always be available at [mixed_input.lua](https://github.com/grayespinoza/REScripts/blob/main/mixed_input.lua).

**Comment by grayespinoza** [comment]
I forgot One Piece of information. As I don't read the controller as K&m when there's input to the right analog, I'm pretty sure conflicts between right analog and mouse gyro should result in only right analog being read.

**Comment by grayespinoza** [comment]
This'll be my last update for the summer.

Unsurprisingly, the RE8 systems work differently and don't want to work at all. It doesn't seem like they like to update frequently or recognize input changes in general. I'm almost certain it has do with how they call input updates. In RE7, I have
```lua
local pad_manager = sdk.get_managed_singleton(sdk.game_namespace("PadManager"))
    local activePad = pad_manager:call("get_activePad")
    if activePad:call("getStickRightVertical") ~= 0 or activePad:call("getStickRightHorizontal") ~= 0 then
      return
    end
    if hid_manager:get_field("<inputMode>k__BackingField") == 0 then
      hid_manager:set_field("<inputMode>k__BackingField", 1)
    end
```
and in RE8
```lua
local pad_manager = sdk.get_managed_singleton(sdk.game_namespace("HIDPadManager"))
    local activePad = pad_manager:call("get_activePad")
    if activePad:call("getStickRightVertical") ~= 0 or activePad:call("getStickRightHorizontal") ~= 0 then
      hid_manager:get_field("<inputMode>k__BackingField"):set_field("V", 0)
      return
    end
    if hid_manager:get_field("<inputMode>k__BackingField"):get_field("V") == 0 then
      hid_manager:get_field("<inputMode>k__BackingField"):set_field("V", 1)
    end
  end
```
You can see what I mean by using the d-pad in both main menus, and then triggering controller input mode by holding the right analog. RE7 works as intended as a button, but RE8 works as a post-toggle. You can tell mixed input does work by having vastly different sensitivities, but it's super inconsistent. I've had it launch into supporting right analog off the bat, but it doesn't stay on after gyro. A way to get it back is to use the right analog during gyro but this is also inconsistent. Overall, this is not a working implementation, just a demonstration of RE8 jank or maybe the implementor's jank.

**Comment by ItsRogueRen** [comment]
Have there been any changes on this front? Started playing PRAGMATA with a Steam Controller only to discover using mouse to aim blocks all controller inputs, making the Steam Controller effectively no different than any normal controller for this game.
